community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/internal/TerminationMark.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright (c) 2002-2016 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.bolt.v1.runtime.internal;
+
+public class TerminationMark
+{
+    private boolean mark = false;
+    private Neo4jError explanation;
+
+    public synchronized void setMark( Neo4jError explanation )
+    {
+        mark = true;
+        this.explanation = explanation;
+    }
+
+    public synchronized boolean get()
+    {
+        return mark;
+    }
+
+    public synchronized Neo4jError explanation()
+    {
+        return explanation;
+    }
+}
```

Comment:
I envision TerminationMark as being used for thread-safe termination, why it always needs to be atomic/synchronized. Using AtomicReference but the responsibility on the using class, while the current implementation solves that for all users.